### DISCUSSION TITLE
Replace history menu with live history search

### DISCUSSION THREAD

**Discussion by jasonvarga** [marlonrichert/zsh-autocomplete#348] [Q&A] [answered]
Is it possible to bind `↑` to live history search instead of the standard history menu?

They seem pretty much the same to me, but being able to use fzf is really nice.

**Comment by marlonrichert** [comment, marked as answer]
Yes, that's possible. You'll have to write some code, but it's fairly trivial:
```zsh
zle -N up-line-or-search
up-line-or-search() {
  if [[ $LBUFFER == *$'\n'* ]] then
    zle up-line
  else
    zle history-incremental-search-backward -w
  fi
}
```

> being able to use fzf is really nice

…except `zsh-autocomplete`'s live history search _does not use `fzf`_ in any way, shape or from. 🙂

**Reply by jasonvarga** [reply]
> …except zsh-autocomplete's live history search does not use fzf 

When I press ctrl+r I see different things depending on whether fzf is installed or not.

Without fzf:

![image](https://user-images.githubusercontent.com/105211/135730553-00f6002b-7ab4-4dc2-95a3-02482cdf52f6.png)

With fzf:

![image](https://user-images.githubusercontent.com/105211/135730566-0887acfa-64e8-4b04-b994-d983b547f34a.png)

**Reply by jasonvarga** [reply]
ctrl+s uses the non-fzf thing though. Do I have something conflicting?

**Reply by marlonrichert** [reply]
If you source `fzf`'s [keybindings.zsh](https://github.com/junegunn/fzf/blob/master/shell/key-bindings.zsh) _after_ sourcing `zsh-autocomplete`, then the former will override the keybindings of the latter.

Ergo:
* If you want `^R` to use `fzf`, then source `zsh-autocomplete` _before_ `keybindings.zsh`.
* If you want `^R` to work the same as `^S`, then source `zsh-autocomplete`_after_ `keybindings.zsh`. Or alternatively, just don't source `keybindings.zsh` at all.

**Reply by jasonvarga** [reply]
Ah, I didn't realize fzf had its own ctrl+r binding. Makes sense.

In that case, I was able to override the up arrow behavior to use fzf history by adding this:

```sh
zle -N up-line-or-search
up-line-or-search() {
  if [[ $LBUFFER == *$'\n'* ]] then
    zle up-line
  else
    fzf-history-widget
  fi
}
```

Thanks for the help!
